src/main/java/net/fabricmc/mappingio/format/ErrorCollector.java

```diff
@@ -0,0 +1,116 @@
+/*
+ * Copyright (c) 2023 FabricMC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package net.fabricmc.mappingio.format;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+
+import org.jetbrains.annotations.ApiStatus;
+
+public interface ErrorCollector {
+	static ErrorCollector create() {
+		return new ErrorCollector() {
+			@Override
+			public void add(Severity severity, String message) throws IOException {
+				errors.add(new ParsingError(severity, message));
+			}
+
+			@Override
+			public List<ParsingError> getErrors() {
+				return errors;
+			}
+
+			private final List<ParsingError> errors = new ArrayList<>();
+		};
+	}
+
+	default void addInfo(String message) throws IOException {
+		add(Severity.INFO, message);
+	}
+
+	default void addWarning(String message) throws IOException {
+		add(Severity.WARNING, message);
+	}
+
+	default void addError(String message) throws IOException {
+		add(Severity.ERROR, message);
+	}
+
+	void add(Severity severity, String message) throws IOException;
+
+	List<ParsingError> getErrors();
```

Comment:
I think we should remove this method: this prevents the interface from being user extendable as users cannot construct ParsingError. If say a user wants to collect warnings or info but fail fast on errors, this method isn't useful either.

`ErrorCollector.create` can expose a different type that exposes a List of collected parsing errors (in a public field or a method)